File: COLLEGE/ASSIGNMENT7/c/Q.8.c
#include <stdio.h>
#include <stdlib.h>

struct Stack {
    int* arr;
    int top;
    int capacity;
};

struct Stack* createStack(int capacity) {
    struct Stack* stack = (struct Stack*)malloc(sizeof(struct Stack));
    stack->capacity = capacity;
    stack->top = -1;
    stack->arr = (int*)malloc(stack->capacity * sizeof(int));
    return stack;
}

int isFull(struct Stack* stack) {
    return stack->top == stack->capacity - 1;
}

int isEmpty(struct Stack* stack) {
    return stack->top == -1;
}

void push(struct Stack* stack, int item) {
    if (isFull(stack)) return;
    stack->arr[++stack->top] = item;
}

int pop(struct Stack* stack) {
    if (isEmpty(stack)) return -1;
    return stack->arr[stack->top--];
}

struct Queue {
    struct Stack* s1;
    struct Stack* s2;
};

struct Queue* createQueue(int capacity) {
    struct Queue* queue = (struct Queue*)malloc(sizeof(struct Queue));
    queue->s1 = createStack(capacity);
    queue->s2 = createStack(capacity);
    return queue;
}

void enqueue(struct Queue* queue, int item) {
    push(queue->s1, item);
}

int dequeue(struct Queue* queue) {
    if (isEmpty(queue->s2)) {
        while (!isEmpty(queue->s1)) {
            push(queue->s2, pop(queue->s1));
        }
    }
    return pop(queue->s2);
}

void display(struct Queue* queue) {
    if (isEmpty(queue->s1) && isEmpty(queue->s2)) {
        printf("Queue is empty\n");
        return;
    }
    printf("Queue elements: ");
    for (int i = 0; i <= queue->s1->top; i++) {
        printf("%d ", queue->s1->arr[i]);
    }
    for (int i = 0; i <= queue->s2->top; i++) {
        printf("%d ", queue->s2->arr[i]);
    }
    printf("\n");
}

int main() {
    struct Queue* queue = createQueue(5);
    enqueue(queue, 10);
    enqueue(queue, 20);
    enqueue(queue, 30);
    printf("Dequeued: %d\n", dequeue(queue));
    display(queue);
    return 0;
}

File: COLLEGE/ASSIGNMENT7/python/Q.8.py
class QueueUsingStacks:
    def __init__(self):
        self.stack1 = []  # Stack for enqueue operations
        self.stack2 = []  # Stack for dequeue operations

    # Function to enqueue an element to the queue
    def enqueue(self, item):
        self.stack1.append(item)  # Push item onto stack1
        print(f"Enqueued {item} to the queue.")

    # Function to dequeue an element from the queue
    def dequeue(self):
        if self.is_empty():
            print("Queue is empty! Cannot dequeue element.")
            return None
        
        # If stack2 is empty, transfer elements from stack1 to stack2
        if not self.stack2:
            while self.stack1:
                self.stack2.append(self.stack1.pop())  # Pop from stack1 and push to stack2
        
        return self.stack2.pop()  # Pop the top element from stack2

    # Function to check if the queue is empty
    def is_empty(self):
        return len(self.stack1) == 0 and len(self.stack2) == 0

    # Function to display the current elements in the queue
    def display(self):
        if self.is_empty():
            print("Queue is empty!")
        else:
            # The current elements in the queue are in stack2 and stack1
            # Elements in stack2 are in correct order, while elements in stack1 need to be reversed
            print("Queue elements:", end=" ")
            temp_stack = self.stack2.copy()  # Copy stack2 to avoid modifying it
            while temp_stack:
                print(temp_stack.pop(), end=" ")
            # Now print stack1 in reverse order
            for item in reversed(self.stack1):
                print(item, end=" ")
            print()  # Newline for better formatting


# Test the QueueUsingStacks class
queue = QueueUsingStacks()

# Enqueue elements to the queue
queue.enqueue(10)
queue.enqueue(20)
queue.enqueue(30)

# Display the queue
queue.display()

# Dequeue elements from the queue
print(f"Dequeued: {queue.dequeue()}")
print(f"Dequeued: {queue.dequeue()}")

# Display the queue after dequeuing
queue.display()

# Enqueue more elements
queue.enqueue(40)
queue.enqueue(50)

# Final state of the queue
queue.display()
